### PR TITLE
Merge "IGAU Codex Data" fork

### PR DESCRIPTION
**Pull request by Elite-IGAU**
All "testing" comments removed
Added "Transmit Codex Discoveries" setting, which now saves correctly. 
If enabled, any "CodexEntry" items in a CMDR journal will be shared to the public IGAU Discovery Catalog. (Future releases may expand this to other Interest Criteria)

HTTP POST function in CodexReader has been cleaned up / optimized. 

Updated SettingsFrm and Observatory.settings to properly "save" the "Transmit Codex Discoveries" setting.

### CODE REVIEW

**Comment by Elite-IGAU** [comment]
Good questions.  
Answers in same order of your questions:

Duplicate submissions are completely ignored. 
More to the point, that's why I use timestamp as the primary index key. 

While it is mathematically possible for several CMDR's to submit data for discoveries all at the EXACT same second, it is very unlikely at the scale we're at.  If this was a stock trading algorithm, then yes, I'd primary key off some other piece of absolutely unique data. 

The backend is API Gateway and DynamoDB, with dynamic scaling policies (albeit with a "handbrake" at a certain capacity level to keep costs reasonable) 

If this is a major concern ( I personally would like to get as much "back catalog" data as we can ), there's no problem doing a minor update, with your guidance to change this code to only ship during "monitoring" which is almost exactly how our EDMC plugin works, and how our upcoming shipper in EDDiscovery will work. 

> 
> Code looks good now, but I have one final concern.
> 
> How do you handle repeat submissions? Because someone who has IGAU submission enabled and hits "Read All Logs" more than once is going to send a _lot_ of them. For that matter, is the endpoint okay with accepting dozens or even hundreds of requests in the span of a couple of seconds when "Read All Logs" gets pressed? Should IGAU submission only ever happen during real-time monitoring?

File: Observatory/CodexReader.cs [hunk 1]
```diff
@@ -3,15 +3,27 @@
 using System.Linq;
 using System.Text;
 using System.Threading.Tasks;
+using System.Net.Http;
```
Comment (Xjph) [inline Observatory/CodexReader.cs]:
```suggestion
using System.Net.Http;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
```

File: Observatory/CodexReader.cs [hunk 2]
```diff
@@ -3,15 +3,27 @@
...
+
     {
         public static void ProcessCodex(CodexEntry codexEntry)
+
         {
-            //Stub method for IGAU Codex Processing
-            
+                  if (Properties.Observatory.Default.SendToIGAU)
+                  {
+                  string POST_content = "{ \"timestamp\":\""+codexEntry.Timestamp.ToString("yyyy-MM-ddTHH:mm:ssZ")+"\", \"Name_Localised\":\""+codexEntry.NameLocalised+"\", \"System\":\""+codexEntry.System+"\" }";
```
Comment (Xjph) [inline Observatory/CodexReader.cs]:
```suggestion
                  JObject POST_content = JObject.FromObject(
                new {
                    timestamp = codexEntry.Timestamp.ToString("yyyy-MM-ddTHH:mm:ssZ"),
                    Name_Localised = codexEntry.NameLocalised,
                    codexEntry.System
                });
```

File: Observatory/CodexReader.cs [hunk 3]
```diff
@@ -3,15 +3,27 @@
...
+                  var request = new System.Net.Http.HttpRequestMessage
+                   {
+                        Method = HttpMethod.Post,
+                        RequestUri = new Uri($"https://ddss70885k.execute-api.us-west-1.amazonaws.com/Prod"),
+                        Content = new StringContent(POST_content)
```
Comment (Xjph) [inline Observatory/CodexReader.cs]:
```suggestion
                        Content = new StringContent(POST_content.ToString(Formatting.None))
```

**Comment by Xjph** [comment]
Made some extra code suggestions in the interest of avoiding writing json by hand.

**Comment by Elite-IGAU** [comment]
Breakpoint in CodexReader.cs provides this output at line 32: 

{Method: POST, RequestUri: 'https://ddss70885k.execute-api.us-west-1.amazonaws.com/Prod', Version: 1.1, Content: System.Net.Http.StringContent, Headers:{  Content-Type: text/plain; charset=utf-8}}

Shouldn't "Content:" be the data to be inserted?

**Comment by Xjph** [comment]
> Breakpoint in CodexReader.cs provides this output at line 32:
> 
> {Method: POST, RequestUri: 'https://ddss70885k.execute-api.us-west-1.amazonaws.com/Prod', Version: 1.1, Content: System.Net.Http.StringContent, Headers:{ Content-Type: text/plain; charset=utf-8}}
> 
> Shouldn't "Content:" be the data to be inserted?

Yes, it's telling you that there is content represented by a StringContent object. It is the json data, you can drill into the object to see it by using the expand buttons in the mouseover tooltips, or by right clicking on the request object and choosing "quickwatch", which will pop up a window that lets you explore it in more detail.

You could also put a line in for testing like `string test = request.Content.ReadAsStringAsync().Result;` to pull the content back out and look at it.

**Comment by Elite-IGAU** [comment]
string test = request.Content.ReadAsStringAsync().Result; Does show the Content data, and it looks correct. 

At one point I was able to get that code block to spit out the returned data from the HTTP POST call. If the data is accepted by API Gateway / Lambda  ( not necessarily written to DynamoDB in case of duplicate data ) it returns "Success"

**Comment by Xjph** [comment]
Yes, you should be able to grab the result back from SendRequest and check it. Is that not working?

**Comment by Elite-IGAU** [comment]
should be able to do something like: 

string test = request.result.ReadAsStringAsync().Result; ??? I think I stumbled upon how to do it, and promptly forgot. :(

**Comment by Xjph** [comment]
Need to grab the response back from HttpClient.SendRequest.
```
HttpResponseMessage response = HttpClient.SendRequest(request);
string responseString = response.Content.ReadAsStringAsync().Result;
```
Or something similar. You can put it all on one line if you like by attaching .Content directly to the SendRequest call, as is done in the version check in ObservatoryFrm.cs. I just wanted to break it out here to explicitly show that you're grabbing .Content from the HttpResponseMessage object that is returned by SendRequest. "HttpResponseMessage" can be replaced with just "var" for its declaration as well, but again, wanted to be explicit about what was happening.

**Comment by Elite-IGAU** [comment]
I think at this point we're good. I'm going to run this build for a few days and do some "live" testing. 

Any additional tweaks/optimizations that you might think of, I'm all ears.

**Pull request merged**